### PR TITLE
front: refacto: use RTK in AddOrEditScenario

### LINKED ISSUE #4158: front: migrate RTK endpoint in AddOrEditScenario.tsx
### Description and goal

Use RTK for endpoints calls

### Acceptance criteria

The endpoints are all called with RTK

### PR DESCRIPTION
**Pull request by Uriel-Sautron**
close #4158

### CODE REVIEW

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx [hunk 1]
```diff
@@ -19,66 +18,83 @@ import { useNavigate } from 'react-router-dom';
...
   const navigate = useNavigate();
   const projectID = useSelector(getProjectID);
   const studyID = useSelector(getStudyID);
   const infraID = useSelector(getInfraID);
-  const [selectedValue, setSelectedValue] = useState(noElectricalProfileSetOption);
+  const [selectedValue, setSelectedValue] = useState<SelectOptionsType | undefined>(
+    noElectricalProfileSetOption
+  );
+
+  type ElectricalProfileSetType = { id: number; name: string };
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx:91]: This can be set up outside the component (to avoid re-creating it at each re-render)

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx [hunk 2]
```diff
@@ -19,66 +18,83 @@ import { useNavigate } from 'react-router-dom';
...
+    osrdEditoastApi.endpoints.patchProjectsByProjectIdStudiesAndStudyIdScenariosScenarioId.useMutation(
+      {}
+    );
+  const [postScenario] =
+    osrdEditoastApi.endpoints.postProjectsByProjectIdStudiesAndStudyIdScenarios.useMutation({});
+
   const noElectricalProfileSetOption = {
     key: undefined,
     value: t('noElectricalProfileSet'),
   };
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: This can be initialized outside of the component

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx [hunk 3]
```diff
@@ -105,48 +119,58 @@ export default function AddOrEditScenarioModal({
...
-        navigate('/operational-studies/scenario');
-        closeModal();
-      } catch (error) {
-        console.error(error);
-      }
+    } else if (infraID && projectID && studyID && currentScenario && infraID) {
+      postScenario({
+        projectId: projectID,
+        studyId: studyID,
+        scenarioRequest: { ...(currentScenario as ScenarioRequest), infra_id: infraID },
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: :warning: you should use the ScenarioRequest from osrdEditoastApi ! (not osrdMiddlewareApi)
```suggestion
        scenarioRequest: currentScenario as ScenarioRequest,
```

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx [hunk 4]
```diff
@@ -105,48 +119,58 @@ export default function AddOrEditScenarioModal({
...
-    } else {
-      try {
-        const result = await post(`${rootURI}`, { ...currentScenario, infra_id: infraID });
-        dispatch(updateScenarioID(result.id));
-        navigate('/operational-studies/scenario');
-        closeModal();
-      } catch (error) {
-        console.error(error);
-      }
+    } else if (infraID && projectID && studyID && currentScenario && infraID) {
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]:
```suggestion
    } else if (projectID && studyID && currentScenario) {
```

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx [hunk 5]
```diff
@@ -105,48 +119,58 @@ export default function AddOrEditScenarioModal({
...
+        scenarioRequest: { ...(currentScenario as ScenarioRequest), infra_id: infraID },
+      })
+        .unwrap()
+        .then(({ id }) => {
+          dispatch(updateScenarioID(id));
+          navigate('/operational-studies/scenario');
+          closeModal();
+        })
+
+        .catch((error) => console.error(error));
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: It would be nice to show a notification to the user here if the request fails

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx [hunk 6]
```diff
@@ -105,48 +119,58 @@ export default function AddOrEditScenarioModal({
...
+    } else if (scenario && projectID && studyID && scenario.id) {
+      patchScenario({
+        projectId: projectID,
+        studyId: studyID,
+        scenarioId: scenario.id,
+        scenarioPatchRequest: currentScenario,
+      })
+        .unwrap()
+        .then(() => {
+          if (getScenarioTimetable) getScenarioTimetable(true);
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx:145]: We should use tags here : if we patch the scenario, then it should unvalidate automatically the scenario and the timetable. Use RTK tags to do this

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx [hunk 7]
```diff
@@ -105,48 +119,58 @@ export default function AddOrEditScenarioModal({
...
+          closeModal();
+        })
+        .catch((error) => console.error(error));
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: If there is an error, we should inform the user

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx [hunk 8]
```diff
@@ -105,48 +119,58 @@ export default function AddOrEditScenarioModal({
...
+          navigate('/operational-studies/study');
+          closeModal();
+          dispatch(
+            setSuccess({
+              title: t('scenarioDeleted'),
+              text: t('scenarioDeletedDetails', { name: scenario.name }),
+            })
+          );
+        })
+        .catch((error) => console.error(error));
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: Same: if the deletion fails, we should inform the user

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx [hunk 9]
```diff
@@ -105,48 +119,58 @@ export default function AddOrEditScenarioModal({
...
-        dispatch(
-          setSuccess({
-            title: t('scenarioDeleted'),
-            text: t('scenarioDeletedDetails', { name: scenario.name }),
-          })
-        );
-      } catch (error) {
-        console.error(error);
-      }
+    if (scenario && projectID && studyID && scenario.id) {
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]:
```suggestion
    if (scenario?.id && projectID && studyID) {
```

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx [hunk 10]
```diff
@@ -19,66 +18,83 @@ import { useNavigate } from 'react-router-dom';
...
   const [currentScenario, setCurrentScenario] = useState<ScenarioListResult>(
     scenario || scenarioTypesDefaults
   );
+  const { data: electricalProfils, isError } = osrdEditoastApi.useGetElectricalProfileSetQuery();
+
   const [displayErrors, setDisplayErrors] = useState(false);
-  const [electricalProfileSetOptions, setElectricalProfileSetOptions] = useState([
-    noElectricalProfileSetOption,
-  ]);
+  const [electricalProfileSetOptions, setElectricalProfileSetOptions] = useState<
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: You should rather use a useMemo to directly define this variable:
```
  const electricalProfileSetOptions = useMemo(() => {
    if (isError || electricalProfils.length < 2) {
      return [];
    }

    const sortedElectricalProfils = sortBy(electricalProfils, ['name']);
    return [
      noElectricalProfileSetOption,
      ...sortedElectricalProfils.map((option: ElectricalProfileSetType) => ({
        key: option.id,
        value: option.name,
      })),
    ];
  }, [electricalProfils]);
```
Try to use RTK to remove the maximum of state values :)

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx [hunk 11]
```diff
@@ -19,66 +18,83 @@ import { useNavigate } from 'react-router-dom';
...
+  >([noElectricalProfileSetOption]);
   const dispatch = useDispatch();
   const navigate = useNavigate();
   const projectID = useSelector(getProjectID);
   const studyID = useSelector(getStudyID);
   const infraID = useSelector(getInfraID);
-  const [selectedValue, setSelectedValue] = useState(noElectricalProfileSetOption);
+  const [selectedValue, setSelectedValue] = useState<SelectOptionsType | undefined>(
+    noElectricalProfileSetOption
+  );
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: Same here, no need for a state value
```suggestion
  const selectedValue: SelectOptionsType = useMemo(
    () =>
      electricalProfileSetOptions.find(
        (option) => option.key === currentScenario.electrical_profile_set_id
      ) || noElectricalProfileSetOption,
    [electricalProfileSetOptions, currentScenario]
  );

```

**Review by clarani — CHANGES_REQUESTED** [review]
Nice job ! I left some comments :)

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx (hunk 2 above)
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx:59]: Done !!

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx (hunk 1 above)
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx:91]: Done !!

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx (hunk 2 above)
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: I need hook useTranslation here

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx (hunk 11 above)
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: Thank you good observation !!

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx (hunk 3 above)
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: Done !!! Thank you

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx (hunk 4 above)
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: Done !!!

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx (hunk 5 above)
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: Done !!!

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx (hunk 8 above)
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: Done !!

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx (hunk 10 above)
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: Done !!

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx (hunk 9 above)
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: Done !!

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx (hunk 7 above)
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: Done !!

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx [hunk 12]
```diff
@@ -9,86 +9,86 @@ import { ModalContext } from 'common/BootstrapSNCF/ModalSNCF/ModalProvider';
...
+
   const noElectricalProfileSetOption = {
     key: undefined,
     value: t('noElectricalProfileSet'),
   };
   const [currentScenario, setCurrentScenario] = useState<ScenarioListResult>(
     scenario || scenarioTypesDefaults
   );
+  const { data: electricalProfils, isError } = osrdEditoastApi.useGetElectricalProfileSetQuery();
+  console.log('electricalProfils:', electricalProfils);
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: Console.log here :)

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx [hunk 13]
```diff
@@ -155,11 +189,6 @@ export default function AddOrEditScenarioModal({
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [infraID]);
 
-  useEffect(() => {
-    createElectricalProfileSetOptions();
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, []);
-
   return electricalProfileSetOptions ? (
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: Can you move the condition ? We need to test if electricalProfileSetOptions is empty or not to display only the dropdown to select the electrical profile
```suggestion
  return (
```

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx [hunk 14]
```diff
@@ -9,86 +9,86 @@ import { ModalContext } from 'common/BootstrapSNCF/ModalSNCF/ModalProvider';
...
+    osrdEditoastApi.usePostProjectsByProjectIdStudiesAndStudyIdScenariosMutation({});
+
   const noElectricalProfileSetOption = {
     key: undefined,
     value: t('noElectricalProfileSet'),
   };
   const [currentScenario, setCurrentScenario] = useState<ScenarioListResult>(
     scenario || scenarioTypesDefaults
   );
+  const { data: electricalProfils, isError } = osrdEditoastApi.useGetElectricalProfileSetQuery();
```
Comment (clarani) [inline front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: Set electricalProfils to [] by default to avoid bugs if there is an error. Also, you can use selectFromResult to remove the variable `electricalProfileSetOptions`
```suggestion
  const { data: electricalProfilOptions = [], isError } = osrdEditoastApi.useGetElectricalProfileSetQuery({
    selectFromResult: (response) => {
      #(code of electricalProfileSetOptions)
    }
  });
```

**Review by clarani — COMMENTED** [review]
Almost there ! :)

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx (hunk 12 above)
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: Oupss

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx (hunk 14 above)
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: Done !!!!

File: front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx (hunk 13 above)
Comment (Uriel-Sautron) [reply front/src/applications/operationalStudies/components/Scenario/AddOrEditScenarioModal.tsx]: Done !!!

**Review by clarani — APPROVED** [review]
LGTM ! Nice job :D

**Review by flomonster — APPROVED** [review]
I only checked the openapi
